### DISCUSSION TITLE
Is it safe to call the Present() function from a non-ui thread

### DISCUSSION THREAD

**Discussion by shi-yan** [gfx-rs/wgpu#2707] [Q&A] [answered]
I know I can create command buffers using threads other than the main thread,

but can I submit the command buffer and call the present() function from non-ui threads ? 

My previous experience with opengl tells me no. But I tried it on linux, it worked. Is it safe to do cross-platform wise?

**Comment by cwfitzgerald** [comment, marked as answer]
Yup, you can call it from any thread.

**Reply by shi-yan** [reply]
I did some tests, 

the behavior doesn't seem to be consistent on different platforms.

on linux, there is no problem.

however, I ran into an issue on Mac. if present_drawable is called from a non-ui thread, the result won't be presented to the window.

**Reply by shi-yan** [reply]
I have written a detailed description of the issue I saw

https://github.com/gfx-rs/wgpu/issues/2727

with a reproduce code.

**Reply by shi-yan** [reply]
so I did more investigation. I read the wgpu-hal code and noticed that Surface has a field "main_thread_id".

In the above code, I created the surface in a non-ui thread. So the "main_thread_id" isn't really the main thread's id.

I adjusted my code to create the Surface on the main thread, and things could render now.

**Reply by cwfitzgerald** [reply]
Interesting, that sounds like it should be part of the contract of create_surface. It makes sense condering the restrictions placed on windows.

**Reply by shi-yan** [reply]
if the raw Metal API doesn't have this restriction (it seems to be the case, as I remember), and this restriction comes from the WGPU layer, then I think this is a bug.

**Reply by cwfitzgerald** [reply]
Alright, could you update the issue description with this information?
